magento/magento-cloud-docker
# xDebug enabled and pre-configured in Developer mode

### Description
This pull request adds Xdebug, by adding another container and allowing nginx to route to it if the xdebug header is present. 

https://github.com/magento/magento-cloud-docker/issues/112

This PR does a couple of things:
1) Adds a container in the docker-compose.yml for xdebug connections (fpm-xdebug)
2) Modifies the Nginx Image config to support switching based on xdebug session cookie
3) One minor config change for xdebug that allows IDEs to connect more easily

This was implemented using the method that was described by di-ivanov on the github issue.



### Fixed Issues (if relevant)
1. https://github.com/magento/magento-cloud-docker/issues/112


### Manual testing scenarios
1. To test #1 you need only include the branch, and rebuild the docker config using the --mode developer flag.
`vendor/bin/ece-docker build:compose --mode developer --sync-engine native  --with-xdebug`
This should then allow you to create the containers using docker-compose. You can connect to the container and see that xdebug is enabled using `php -v`, also confirm it is not in the fpm container.

2. To test #2 from the list above, It is a little trickier. You will have to add a docker-compose.override.yml as follows:
```
version: '2.1'
services:
  web:
    build:
      context: ./vendor/magento/magento-cloud-docker/images/nginx/1.9/

```
Then you can use `docker-compose up -d --rebuild` to rebuild the container. This is only needed until nginx is published, but to test the nginx you will need the config.
Once you have done that, you should be able to use xdebug as normal, if you have the cookie set you will see traffic pass to the xdebug container.

The following chrome extension adds the proper header:
https://chrome.google.com/webstore/detail/xdebug-helper/eadndfjplgieldjbigjakmdgkmoaaaoc

Please let me know if you have any questions, I know this is a bit harder than normal to test.


### Contribution checklist
